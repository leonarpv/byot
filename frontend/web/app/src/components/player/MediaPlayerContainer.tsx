import * as React from 'react';
import {MediaPlayerSkeleton} from './MediaPlayerSkeleton';
import {makeStyles, Theme} from '@material-ui/core';

interface Props {
  children?: React.ReactNode;
  ratio?: '16:9' | '4:3';
}

const styles = (theme: Theme) => ({
  root: {
    width: '100%',
  },
  skeleton: {
    height: theme.spacing(20),
  },
  container: {
    width: '100%',
    overflow: 'hidden',
    borderRadius: theme.shape.borderRadius,
  },
});
const useStyles = makeStyles(styles);

export const MediaPlayerContainer: React.FC<Props> = (props: Props) => {
  const styles = useStyles(props);
  const [height, setHeight] = React.useState(0);
  const [node, setNode] = React.useState<HTMLDivElement | undefined>(undefined);
  const ref = React.useCallback(node => setNode(node), []);
  const ratio = props.ratio || '16:9';
  React.useLayoutEffect(() => {
    const onResize = () => {
      const width = node?.getBoundingClientRect().width || 0;
      if (width > 0) {
        const [w, h] = ratio.split(':');
        setHeight((width / parseInt(w)) * parseInt(h));
      }
    };
    onResize();
    window.addEventListener('resize', onResize);
    return () => window.removeEventListener('resize', onResize);
  }, [node, ratio]);
  return (
    <div ref={ref} className={styles.root}>
      {height == 0 ? (
        <div className={styles.skeleton}>
          <MediaPlayerSkeleton />
        </div>
      ) : (
        <div
          className={styles.container}
          style={{height: `${height}px`}}
          data-testid="media-player-container">
          {props.children}
        </div>
      )}
    </div>
  );
};
