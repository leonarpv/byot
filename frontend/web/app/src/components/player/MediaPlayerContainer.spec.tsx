import * as React from 'react';
import {render, waitFor, act, fireEvent} from '@testing-library/react';
import {MediaPlayerContainer} from './MediaPlayerContainer';

describe('<MediaPlayerContainer/>', () => {
  it('should render', () => {
    const {container} = render(<MediaPlayerContainer />);
    expect(container).toMatchSnapshot();
  });
  it('should calculate width (16:9)', async () => {
    jest.spyOn(HTMLDivElement.prototype, 'getBoundingClientRect').mockImplementation(
      () =>
        ({
          width: 1600,
        } as any)
    );
    const {container, getByTestId} = render(<MediaPlayerContainer />);
    const mpContainer = getByTestId('media-player-container');
    expect(mpContainer.style.height).toEqual('900px');
    expect(container).toMatchSnapshot();
  });
  it('should calculate width (4:3)', async () => {
    jest.spyOn(HTMLDivElement.prototype, 'getBoundingClientRect').mockImplementation(
      () =>
        ({
          width: 400,
        } as any)
    );
    const {container, getByTestId} = render(<MediaPlayerContainer ratio="4:3" />);
    const mpContainer = getByTestId('media-player-container');
    expect(mpContainer.style.height).toEqual('300px');
    expect(container).toMatchSnapshot();
  });
});
