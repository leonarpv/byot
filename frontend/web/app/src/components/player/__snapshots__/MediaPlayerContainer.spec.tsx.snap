// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaPlayerContainer/> should calculate width (4:3) 1`] = `
<div>
  <div
    class="makeStyles-root-23"
  >
    <div
      class="makeStyles-container-25"
      data-testid="media-player-container"
      style="height: 300px;"
    />
  </div>
</div>
`;

exports[`<MediaPlayerContainer/> should calculate width (16:9) 1`] = `
<div>
  <div
    class="makeStyles-root-12"
  >
    <div
      class="makeStyles-container-14"
      data-testid="media-player-container"
      style="height: 900px;"
    />
  </div>
</div>
`;

exports[`<MediaPlayerContainer/> should render 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="makeStyles-skeleton-2"
    >
      <div
        class="MuiBox-root MuiBox-root-10 makeStyles-root-4 makeStyles-root-9 makeStyles-root-4"
        classes="[object Object]"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          >
            <span
              class="MuiSkeleton-root makeStyles-root-11 MuiSkeleton-rect MuiSkeleton-pulse"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          >
            <span
              class="MuiSkeleton-root makeStyles-root-11 MuiSkeleton-rect MuiSkeleton-pulse"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <span
              class="MuiSkeleton-root makeStyles-root-11 MuiSkeleton-rect MuiSkeleton-pulse"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          >
            <span
              class="MuiSkeleton-root makeStyles-root-11 MuiSkeleton-rect MuiSkeleton-pulse"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
